tls/s2n_ktls.c

```diff
@@ -0,0 +1,203 @@
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE(cipher->ktls_supported, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_validate_socket_mode(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE(ktls_mode == S2N_KTLS_MODE_RECV || ktls_mode == S2N_KTLS_MODE_SEND, S2N_ERR_SAFETY);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     *
+     * - Safety checks
+     * - Confirm application is not using custom I/O
+     * - Confirm kTLS isn't enabled already
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE(conn->managed_send_io, S2N_ERR_KTLS_SEND_MANAGED_IO);
+            RESULT_ENSURE(!conn->ktls_send_enabled, S2N_ERR_KTLS_SEND_ENABLED);
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE(conn->managed_recv_io, S2N_ERR_KTLS_RECV_MANAGED_IO);
+            RESULT_ENSURE(!conn->ktls_recv_enabled, S2N_ERR_KTLS_RECV_ENABLED);
+            break;
+        case S2N_KTLS_MODE_DUPLEX:
+            RESULT_BAIL(S2N_ERR_SAFETY);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        RESULT_ENSURE_REF(conn->recv_io_context);
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        RESULT_ENSURE_REF(conn->send_io_context);
+        /* retrieve the send fd */
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+
+#ifdef S2N_PLATFORM_SUPPORTS_KTLS
+    /* Enable 'tls' ULP for the socket. https://lwn.net/Articles/730207 */
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
+    if (ignore_ktls_ulp_for_testing == false) {
+        int ret = setsockopt(fd, SOL_TCP, TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE);
+        RESULT_ENSURE(ret > S2N_FAILURE, S2N_ERR_KTLS_ULP);
```

Comment:
We generally don't compare non-s2n return values to S2N_FAILURE. I'd write out "-1". But based on the setsockopt documentation, maybe you should be checking for "==0".

> On success, zero is returned. On error, -1 is returned, and errno is set appropriately.